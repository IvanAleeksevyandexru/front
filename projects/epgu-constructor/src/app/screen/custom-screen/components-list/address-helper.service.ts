import { Injectable } from '@angular/core';
import { Observable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { DadataSuggestions, DadataSuggestionsAddress } from '../../../services/api/dadata-api/dadata-api.types';
import { DictionaryApiService } from '../../../services/api/dictionary-api/dictionary-api.service';

export interface DadataSuggestionsAddressForLookup extends DadataSuggestionsAddress {
  id: string,
  text: string,
}

interface objectWithSuggestions {
  suggestions: DadataSuggestions
}

@Injectable()
export class AddressHelperService {

  constructor(private dictionaryApiService: DictionaryApiService) { }

  // Провайдер поиска для передачи в lib-lookup
  // с функцией поиска для lib-lookup. Сам поиск осуществляется за счет suggestions дадаты
  public provider = { search: (searchString) => searchString ? this.getCitySuggestions(searchString) : of([]) };

  /**
   * Получение городов из suggestions дадаты для lib-lookup. Добавляет к suggestions атрибуты id и text
   * @param qString - строка для поиска
   */
  public getCitySuggestions(qString: string): Observable<Array<DadataSuggestionsAddressForLookup>> {
    return this.dictionaryApiService.getDadataSuggestions(qString, { isCity: 'true' })
      .pipe(
        map(({ suggestions }: objectWithSuggestions) => {
          return suggestions.addresses
            .map((address) => {
              return {
                ...address,
                id: address.code,
                text: address.address,
              };
            });
        }),
      );
  }

  /**
   * "Наращивает" адресс из suggestions дадаты данными из normalize
   * @param address - объект с адресом из suggestions
   */
  public async normalizeAddress(address: DadataSuggestionsAddressForLookup) {
    const normalAddress = await this.dictionaryApiService.getDadataNormalize(address.address).toPromise();
    Object.assign(address, normalAddress);
  }
}
