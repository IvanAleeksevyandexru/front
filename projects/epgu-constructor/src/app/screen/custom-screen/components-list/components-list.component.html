<form #customForm="ngForm" action="" class="form">
  <ng-container *ngFor="let componentData of components">
    <ng-container [ngSwitch]="componentData.type">
      <ng-container *ngSwitchCase="componentType.LabelSection">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-output-html
            class="label"
            [html]="componentData.label || state[componentData.id].value"
            [clarifications]="componentData?.attrs.clarifications"
          >
          </epgu-constructor-output-html>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.Dictionary">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <lib-dropdown
            class="cs__select cs-select"
            [disabled]="state[componentData.id].disabled"
            [items]="dictionary[componentData?.attrs?.dictionaryType + componentData?.id]?.list"
            (changed)="selectDictionary($event, componentData)"
          ></lib-dropdown>
          <epgu-constructor-helper-text *ngIf="!componentData.required">
            {{ getHelperText(componentData.required) }}
          </epgu-constructor-helper-text>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.GenderSelection">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <epgu-constructor-gender-radio-button
            [disabled]="state[componentData.id].disabled"
            [(ngModel)]="state[componentData.id].value">
          </epgu-constructor-gender-radio-button>
          <epgu-constructor-helper-text *ngIf="!componentData.required">
            {{ getHelperText(componentData.required) }}
          </epgu-constructor-helper-text>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.DropDown">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <lib-dropdown
            class="cs__select cs-select"
            [disabled]="state[componentData.id].disabled"
            [items]="dropDown[componentData?.id]?.list"
            (changed)="selectDropDown($event, componentData)"
          ></lib-dropdown>
          <epgu-constructor-helper-text *ngIf="!componentData.required">
            {{ getHelperText(componentData.required) }}
          </epgu-constructor-helper-text>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.Lookup">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <lib-lookup
            [queryMinSymbolsCount]="0"
            [searchCaseSensitive]="false"
            [disabled]="state[componentData.id].disabled"
            [fixedItems]="
              dictionary[componentData?.attrs?.dictionaryType + componentData?.id]?.list
            "
            (changed)="selectDictionary($event, componentData)"
          ></lib-lookup>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.htmlString">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-output-html
            *ngIf="!state[componentData.id].disabled"
            class="info__text"
            [html]="componentData?.label"
            [clarifications]="componentData?.attrs?.clarifications"
          >
          </epgu-constructor-output-html>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.StringInput">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <lib-standard-masked-input
            *ngIf="componentData.attrs?.mask; else plainInput"
            [clearable]="true"
            [mask]="componentData.attrs?.mask | maskHandle"
            [placeholder]="componentData.attrs?.placeholder"
            [name]="componentData.id"
            [disabled]="state[componentData.id].disabled"
            [showMaskAsPlaceholder]="false"
            [showConstantMaskSymbols]="true"
            [(ngModel)]="state[componentData.id].value"
            (change)="inputChange($event, componentData)"
          ></lib-standard-masked-input>
          <ng-template #plainInput>
            <lib-plain-input
              epguConstructorTrim
              [invalid]="!state[componentData.id].valid"
              [name]="componentData.id"
              [disabled]="state[componentData.id].disabled"
              [(ngModel)]="state[componentData.id].value"
              (change)="inputChange($event, componentData)"
              [validationShowOn]="validationShowOn"
            ></lib-plain-input>
          </ng-template>
          <div *ngIf="componentData.attrs?.hint" class="form__item__hint">
            {{ componentData.attrs?.hint }}
          </div>
          <epgu-constructor-helper-text *ngIf="!componentData.required">
            {{ getHelperText(componentData.required) }}
          </epgu-constructor-helper-text>
          <div class="form__error">
            <small *ngIf="!state[componentData.id].valid"
            >{{ state[componentData.id]?.errorMessage }}
            </small>
          </div>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.CheckBox">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <div [ngClass]="
              componentData?.attrs?.isHorizontal ? 'form__radio_inline' : 'form__radio_block'
            ">
            <lib-checkbox
              [checkboxId]="componentData.id"
              [checked]="componentData?.attrs?.checked"
              [disabled]="state[componentData.id].disabled"
              [labelText]="componentData?.label"
              (change)="checkboxFieldsTogglerChange($event, componentData)"
            >
            </lib-checkbox>
          </div>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.RadioInput">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <div
            [ngClass]="
              componentData?.attrs?.isHorizontal ? 'form__radio_inline' : 'form__radio_block'
            "
          >
            <lib-radio
              *ngFor="let item of componentData?.attrs?.supportedValues"
              [name]="componentData.id"
              [labelText]="item.label"
              [disabled]="state[componentData.id].disabled"
              [value]="item.value"
              [checked]="item.isDefault"
              (change)="inputChange($event, componentData)"
            >
            </lib-radio>
          </div>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.DateInput">
        <div class="form__item form__item--date" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <!-- нужно добавить changes для проверки зависимости от даты-->
          <lib-date-picker
            [name]="componentData.id"
            [disabled]="state[componentData.id].disabled"
            (ngModelChange)="dateChange($event, componentData)"
            [(ngModel)]="state[componentData.id].value"
            minDate="-18y"
          ></lib-date-picker>
        </div>
      </ng-container>

      <ng-container *ngSwitchCase="componentType.AddressInput">
        <div class="form__item" *ngIf="state[componentData.id].isShown">
          <epgu-constructor-label>{{ componentData.label }}</epgu-constructor-label>
          <lib-dadata-widget
            [simpleMode]="true"
            [name]="componentData.id"
            [hideLevels]="componentData?.attrs?.hideLevels || []"
            [disabled]="state[componentData.id].disabled"
            [externalApiUrl]="config.externalApiUrl + '/'"
            [initValue]="state[componentData.id].value"
            [(ngModel)]="state[componentData.id].value"
            (change)="inputChange($event, componentData)"
          >
          </lib-dadata-widget>
        </div>
      </ng-container>
    </ng-container>
  </ng-container>
</form>
