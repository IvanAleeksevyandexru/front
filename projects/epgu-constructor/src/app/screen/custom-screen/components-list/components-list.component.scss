@import '../../../../styles/core/palettes';
@import '../../../../styles/core/typography';
@import '../../../../styles/core/variables';

// test
.form {
  &__item {
    margin-bottom: $screen-pad-p;

    &__hint {
      margin-top: 4px;
      @include typ-plain-xs();
    }
  }
  &__title {
    font-size: 14px;
    font-family: $font-main;
    line-height: 20px;
    color: $color-summer-night;
    margin-bottom: 4px;
  }
  &__error {
    margin-top: 4px;
    color: $color-gosred;
  }
  &__radio {

    &__block {
      margin-top: $screen-pad-p;
      display: inline-block;
      lib-radio {
        display: block;
        margin-bottom: 16px;
      }
    }
    &__inline {
      margin-top: $screen-pad-p;
      display: inline-flex;
      lib-radio:not(:first-child) {
        margin-left: $screen-pad-p;
      }
    }
  }
}

.label {
  margin-bottom: $screen-pad-p;
  font-size: 16px;
  line-height: 24px;
  font-weight: bold;
}
