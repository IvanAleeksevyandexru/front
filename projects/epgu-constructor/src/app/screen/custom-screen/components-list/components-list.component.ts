import {
  Component,
  EventEmitter,
  Input,
  OnChanges,
  OnInit,
  Output,
  SimpleChanges,
} from '@angular/core';
import { ListItem, ValidationShowOn } from 'epgu-lib';
import { ConfigService } from '../../../config/config.service';
import { DictionaryApiService } from '../../../services/api/dictionary-api/dictionary-api.service';
import { DictionaryResponse } from '../../../services/api/dictionary-api/dictionary-api.types';
import { OPTIONAL_FIELD } from '../../../shared/constants/helper-texts';
import { ScreenService } from '../../screen.service';
import {
  CustomComponent,
  CustomComponentDictionaryState,
  CustomComponentDropDownStateInterface,
  CustomComponentOutputData,
  CustomComponentState,
  CustomScreenComponentTypes,
} from '../custom-screen.types';
import {
  adaptiveDropDown,
  calcDependedComponent,
  checkLegalInn,
  CheckInputValidationComponentList,
  getCustomScreenDictionaryFirstState,
  getInitStateItemComponentList,
  getNormalizeDataCustomScreenDictionary,
  isDropDown,
  likeDictionary,
  checkPersonInn,
} from '../tools/custom-screen-tools';

@Component({
  selector: 'epgu-constructor-components-list',
  templateUrl: './components-list.component.html',
  styleUrls: ['./components-list.component.scss'],
})
export class ComponentsListComponent implements OnInit, OnChanges {
  // <-- constant
  componentType = CustomScreenComponentTypes;

  // <-- variables
  validationShowOn = ValidationShowOn.TOUCHED_UNFOCUSED;
  state: CustomComponentState = {};
  dictionary: { [key: string]: CustomComponentDictionaryState } = {};
  dropDown: { [key: string]: CustomComponentDropDownStateInterface } = {};

  @Input() components: Array<CustomComponent>;
  @Output() changes = new EventEmitter<CustomComponentOutputData>();

  constructor(
    private dictionaryApiService: DictionaryApiService,
    public screenService: ScreenService,
    public config: ConfigService,
  ) {}

  // NOTICE: тут была информация о валидации смотри историю гита

  ngOnChanges(changes: SimpleChanges): void {
    this.state = {};
    if (changes?.components?.currentValue) {
      this.components.forEach((component) => this.initComponent(component));
      this.emmitChanges();
      this.checkDependenceOfTheComponent();
    }
  }

  /**
   * Инициализирует стейт для компонента.
   */
  initComponent(component: CustomComponent) {
    this.initState(component);
    if (likeDictionary(component.type)) {
      const { dictionaryType } = component.attrs;
      this.initDictionary(dictionaryType, component.id);
      this.loadDictionary(dictionaryType, component);
    } else if (isDropDown(component.type)) {
      this.initDropDown(component);
    }
  }

  initState(component: CustomComponent) {
    this.state[component.id] = getInitStateItemComponentList(component);
  }

  initDictionary(dictionaryType, componentId) {
    this.dictionary[dictionaryType + componentId] = getCustomScreenDictionaryFirstState();
  }

  initDropDown(component: CustomComponent) {
    const key = component.id;
    const data = component.attrs.dictionaryList;
    this.dropDown[key] = {
      origin: data,
      list: adaptiveDropDown(data),
    };
  }

  selectDictionary(selectedItem: ListItem, component: CustomComponent) {
    const dictionaryType = component.attrs?.dictionaryType;
    this.dictionary[dictionaryType + component.id].selectedItem = selectedItem.originalItem;
    this.state[component.id].value = selectedItem.originalItem;
    this.state[component.id].valid = true;
    this.emmitChanges(component);
  }

  selectDropDown($event: any, componentData: CustomComponent) {
    this.state[componentData.id].value = $event.origin;
    this.emmitChanges();
  }

  inputChange($event: Event, component: CustomComponent) {
    const { value } = $event.target as HTMLInputElement;
    this.state[component.id].value = value;
    const inputValidationResult = CheckInputValidationComponentList(value, component);
    this.setValidationState(inputValidationResult, component.id, value);
    this.emmitChanges(component);
  }

  dateChange($event: string, component: CustomComponent) {
    const value = $event;
    this.state[component.id].value = value;
    const inputValidationResult = CheckInputValidationComponentList(value, component);
    this.setValidationState(inputValidationResult, component.id, value);
    this.emmitChanges(component);
  }

  loadDictionary(dictionaryType: string, component: CustomComponent) {
    // TODO добавить обработку loader(-а) для словарей и ошибок;
    this.dictionaryApiService.getDictionary(dictionaryType, { pageNum: 0 }).subscribe(
      (data) => this.loadDictionarySuccess(dictionaryType, data, component),
      () => this.loadDictionaryError(dictionaryType, component.id),
      () => {
        /* this.changes.emit(this.state) */
      },
    );
  }

  loadDictionarySuccess(key: string, data: DictionaryResponse, component: CustomComponent) {
    const id = key + component.id;
    this.dictionary[id].loading = false;
    this.dictionary[id].paginationLoading = false;
    this.dictionary[id].data = data;
    this.dictionary[id].origin = component;
    this.dictionary[id].list = getNormalizeDataCustomScreenDictionary(data.items, key, component);
  }

  loadDictionaryError(key: string, componentId: string) {
    const id = key + componentId;
    this.dictionary[id].loading = false;
    this.dictionary[id].paginationLoading = false;
    this.dictionary[id].loadError = true;
    this.dictionary[id].loadEnd = false;
  }

  setValidationState(inputValidationResult, componentId, componentValue) {
    const handleSetState = (isValid, errMsg?) => {
      this.state[componentId].value = componentValue;
      this.state[componentId].valid = isValid;
      this.state[componentId].errorMessage = errMsg;
    };

    if (inputValidationResult === -1) {
      handleSetState(true);
    } else {
      handleSetState(
        false,
        this.state[componentId]?.component.attrs.validation[inputValidationResult].errorMsg,
      );
    }
  }

  getHelperText(required: boolean): string {
    return required ? '' : OPTIONAL_FIELD;
  }

  emmitChanges(component?: CustomComponent) {
    if (component) {
      calcDependedComponent(component, this.state, this.components);
    }
    const prepareStateForSending = this.getPreparedStateForSending();
    this.changes.emit(prepareStateForSending);
  }

  legalInnInputChange($event: Event, component: CustomComponent) {
    const { value } = $event.target as HTMLInputElement;
    this.state[component.id].value = value;
    const inputValidationResult = checkLegalInn(value);
    this.setValidationState(inputValidationResult, component.id, value);
    this.emmitChanges(component);
  }

  personInnInputChange($event: Event, component: CustomComponent) {
    const { value } = $event.target as HTMLInputElement;
    this.state[component.id].value = value;
    const inputValidationResult = checkPersonInn(value);
    this.setValidationState(inputValidationResult, component.id, value);
    this.emmitChanges(component);
  }

  /**
   * Подгатавливаются данные для пробраса на верх.
   * Родительскому компоненту нужны данные компонента и состояние валидности.
   */
  private getPreparedStateForSending() {
    return Object.entries(this.state).reduce((acc, [key, val]) => {
      const { value, valid, isShown } = val;
      if (isShown) {
        acc[key] = { value, valid };
      }
      return acc;
    }, {});
  }

  private checkDependenceOfTheComponent() {
    this.components.forEach((component) =>
      calcDependedComponent(component, this.state, this.components),
    );
  }

  ngOnInit(): void {}
}
