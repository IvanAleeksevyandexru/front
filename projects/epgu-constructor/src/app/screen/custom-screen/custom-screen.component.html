<epgu-constructor-screen-container
  [show-nav]="screenService.showNav$ | async"
  [class.screen-container-mt]="!(screenService.showNav$ | async)"
>
  <div class="screen-body">
    <epgu-constructor-page-name>{{ screenService.header$ | async }}</epgu-constructor-page-name>

    <epgu-constructor-screen-pad class="body">
      <epgu-constructor-components-list
        [components]="$any((screenService.display$ | async)?.components)"
        [errors]="screenService.componentErrors$ | async"
        (changes)="changeComponentsList($event)"
      >
      </epgu-constructor-components-list>
    </epgu-constructor-screen-pad>
  </div>

  <div class="screen-footer">
    <!-- TODO: добавить поддержку вывода ScreenActionDto  -->
    <lib-button
      *ngIf="screenService.submitLabel$ | async"
      [showLoader]="screenService.isLoading$ | async"
      epgu-constructor-action
      [action]="nextStepAction"
      class="submit-button"
      [disabled]="(screenService.isLoading$ | async) || !isValid"
    >
      {{ screenService.submitLabel$ | async }}
    </lib-button>
  </div>
</epgu-constructor-screen-container>
