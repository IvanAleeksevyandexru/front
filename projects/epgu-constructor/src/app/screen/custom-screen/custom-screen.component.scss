@import "../../../styles/core";

.body {
  display: flex;
  text-align: left;
  @include tablets-down {
    flex-grow: 1;
  }
  > * {
    display: flex;
    width: 100%;
    flex-direction: column;
  }
}
