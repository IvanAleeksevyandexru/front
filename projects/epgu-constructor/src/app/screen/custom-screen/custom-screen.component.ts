import { Component, OnChanges, OnInit } from '@angular/core';
import * as moment_ from 'moment';
import { takeUntil } from 'rxjs/operators';
import { NavigationPayload } from '../../form-player.types';
import { UnsubscribeService } from '../../services/unsubscribe/unsubscribe.service';
import { DATE_STRING_DOT_FORMAT } from '../../shared/constants/dates';
import { NavigationService } from '../../shared/services/navigation/navigation.service';
import { ScreenService } from '../screen.service';
import { Screen, ScreenStore } from '../screen.types';

const moment = moment_;

@Component({
  selector: 'epgu-constructor-custom-screen',
  templateUrl: './custom-screen.component.html',
  styleUrls: ['./custom-screen.component.scss'],
  providers: [UnsubscribeService],
})
export class CustomScreenComponent implements OnInit, OnChanges, Screen {
  dataToSend: NavigationPayload;
  isCycledFields: boolean;
  cycledValues: any;
  screenStore: ScreenStore;
  valid: boolean;

  private currentCycledFields = this.screenStore?.currentCycledFields || {};
  private cycledFieldsKeys = Object.keys(this.currentCycledFields);

  constructor(
    private navigationService: NavigationService,
    private ngUnsubscribe$: UnsubscribeService,
    public screenService: ScreenService,
  ) {}

  ngOnInit(): void {
    this.navigationService.clickToBack$
      .pipe(takeUntil(this.ngUnsubscribe$))
      .subscribe(() => this.prevStep());

    this.screenService.screenData$
      .pipe(takeUntil(this.ngUnsubscribe$))
      .subscribe((screenData: ScreenStore) => {
        this.screenStore = screenData;
        this.initCycledFields();
      });
  }

  ngOnChanges(changes) {
    const { firstChange, currentValue, previousValue } = changes.data;
    if (firstChange || currentValue.id !== previousValue.id) {
      this.initCycledFields();
    }
  }

  initCycledFields() {
    this.currentCycledFields = this.screenStore?.currentCycledFields || {};
    this.cycledFieldsKeys = Object.keys(this.currentCycledFields);

    const { currentCycledFields } = this;
    this.isCycledFields = !!Object.keys(currentCycledFields).length;
    if (this.isCycledFields) {
      const [firstCurrentCycledValues] = Object.values(currentCycledFields);
      this.cycledValues = JSON.parse(firstCurrentCycledValues);
      this.screenStore.display.components = this.screenStore.display.components.map((item) => {
        const newItem = item;
        const fieldName = item.attrs?.fields && item.attrs?.fields[0].fieldName;
        const cycledFieldKey = Object.keys(this.cycledValues).find((key) => key === fieldName);
        newItem.value = this.cycledValues[cycledFieldKey];
        return newItem;
      });
    }
  }

  prevStep(): void {
    this.navigationService.prevStep.next();
  }

  nextStep(payload?: NavigationPayload): void {
    this.navigationService.nextStep.next({ payload });
  }

  nextScreen(): void {
    const data = this.dataToSend;
    this.nextStep(data);
  }

  /**
   * Форматиркем данные перед отправкой
   * @param changes - данные на отправку
   */
  getFormattedData(changes) {
    let stateData = {};
    if (this.isCycledFields) {
      const [currentCycledFieldsKey] = this.cycledFieldsKeys;
      const stateDataPrepared = this.formatCycledFieldsValues(changes);
      const cycledValuesPrepared = { ...this.cycledValues };
      const mergedCycledAndStateData = { ...cycledValuesPrepared, ...stateDataPrepared };

      stateData[currentCycledFieldsKey] = {
        visited: true,
        value: JSON.stringify(mergedCycledAndStateData),
      };
    } else {
      stateData = this.getPrepareResponseData(changes);
    }

    return stateData;
  }

  changeComponentsList(changes: { [key: string]: any }): void {
    if (Object.values(changes).every((item) => item.valid)) {
      this.dataToSend = this.getFormattedData(changes);
      this.valid = true;
    } else {
      this.valid = false;
    }
  }

  /**
   * Возвращает true, если дата валидна
   * @param date - дата для проверки
   * @private
   */
  private isValidDate(date): boolean {
    return new Date(date).toString() !== 'Invalid Date';
  }

  /**
   * Подготавливаем данные для ответа
   * @param data - данные для пребразования
   * @private
   */
  private getPrepareResponseData(data = {}) {
    return Object.keys(data).reduce((acc, key) => {
      let value = '';
      const dataValue = data[key].value;

      if (typeof dataValue === 'object') {
        if (this.isValidDate(dataValue)) {
          value = moment(dataValue).toISOString();
        } else {
          value = JSON.stringify(dataValue || {});
        }
      } else {
        value = dataValue;
      }

      acc[key] = {
        visited: true,
        value,
        disabled: data[key].disabled,
      };

      return acc;
    }, {});
  }

  /**
   * Форматирует значения полей с вложенностью
   * @param changes - данные
   * @private
   */
  private formatCycledFieldsValues(changes: any) {
    return Object.keys(changes).reduce((accum, key) => {
      const result = accum;
      const targetItem = changes[key];
      const targetItemValue = targetItem.value;
      const targetComponent = this.screenStore.display.components.find((item) => item.id === key);
      const fieldName = targetComponent.attrs.fields && targetComponent.attrs.fields[0].fieldName;
      if (!fieldName) return result;

      if (targetComponent.type === 'date') {
        result[fieldName] = moment(targetItemValue).format(DATE_STRING_DOT_FORMAT);
      } else {
        result[fieldName] = targetItemValue;
      }
      return result;
    }, {});
  }
}
