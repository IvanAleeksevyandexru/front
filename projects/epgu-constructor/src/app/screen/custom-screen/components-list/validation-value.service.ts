import { Injectable } from '@angular/core';
import { checkOgrn } from 'ru-validation-codes';
import { CustomScreenComponentTypes } from '../custom-screen.types';

export class ValidationValueService {
  customcomponentType = CustomScreenComponentTypes;
  isValueValid(type, value): boolean {
    if (type === this.customcomponentType.OgrnInput) {
      return checkOgrn(value);
    }
  }
}
