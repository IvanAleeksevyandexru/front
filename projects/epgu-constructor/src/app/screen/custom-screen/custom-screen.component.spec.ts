import { ComponentFixture, TestBed } from '@angular/core/testing';
import { EpguLibModule } from 'epgu-lib';
import { MockComponent } from 'ng-mocks';
import { UnsubscribeService } from '../../services/unsubscribe/unsubscribe.service';
import { PageNameComponent } from '../../shared/components/base/page-name/page-name.component';
import { NavigationComponent } from '../../shared/components/navigation/navigation.component';
import { ScreenContainerComponent } from '../../shared/components/screen-container/screen-container.component';
import { ScreenPadComponent } from '../../shared/components/screen-pad/screen-pad.component';
import { NavigationService } from '../../shared/services/navigation/navigation.service';
import { ScreenService } from '../screen.service';
import { ScreenStore, ScreenTypes } from '../screen.types';
import { ComponentsListComponent } from './components-list/components-list.component';
import { CustomScreenComponent } from './custom-screen.component';

xdescribe('CustomScreenComponent', () => {
  let component: CustomScreenComponent;
  let fixture: ComponentFixture<CustomScreenComponent>;
  let navigationService: NavigationService;
  let screenService: ScreenService;
  let unsubscribeService: UnsubscribeService;
  let NavigationComponentMock = MockComponent(NavigationComponent);
  let ComponentsListComponentMock = MockComponent(ComponentsListComponent);
  const screenDataMock: ScreenStore = {
    display: {
      components: [],
      header: '',
      id: '',
      name: '',
      submitLabel: '',
      type: ScreenTypes.QUESTION
    }
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        EpguLibModule,
      ],
      declarations: [
        CustomScreenComponent,
        PageNameComponent,
        ScreenPadComponent,
        ScreenContainerComponent,
        NavigationComponentMock,
        ComponentsListComponentMock
      ],
      providers: [
        NavigationService,
        UnsubscribeService,
        ScreenService,
      ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CustomScreenComponent);
    component = fixture.componentInstance;

    navigationService = fixture.debugElement.injector.get(NavigationService);
    unsubscribeService = fixture.debugElement.injector.get(UnsubscribeService);
    screenService = fixture.debugElement.injector.get(ScreenService);

    screenService.updateScreenStore(screenDataMock);
    component.changeComponentsList({});
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('check snapshot', () => {
    // TODO: Нужен рефакторинг, т.к. тест не проходит
    //expect(fixture).toMatchSnapshot();
  });

  describe('navigation cases', () => {
    it('onClick lib button should call nextScreen()', () => {
      spyOn(component, 'nextScreen').and.callThrough();
      const button = fixture.debugElement.nativeElement.querySelector('.footer__btn-submit .button');
      button.click();
      fixture.detectChanges();
      expect(component.nextScreen).toHaveBeenCalled();
    });

    it('nextScreen() should call next of nextStep subject from navigationService', () => {
      spyOn(navigationService.nextStep, 'next').and.callThrough();
      component.nextScreen();
      expect(navigationService.nextStep.next).toHaveBeenCalled();
    });

    it('prevStep() should call next of prevStep subject from navigationService', () => {
      spyOn(navigationService.prevStep, 'next').and.callThrough();
      component.prevStep();
      expect(navigationService.prevStep.next).toHaveBeenCalled();
    });
  });
});
