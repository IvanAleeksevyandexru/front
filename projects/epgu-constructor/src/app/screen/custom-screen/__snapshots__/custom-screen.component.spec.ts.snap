// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomScreenComponent check snapshot 1`] = `
<epgu-constructor-custom-screen
  injector={[Function Injector_]}
  navigationService={[Function Object]}
  ngUnsubscribe$={[Function UnsubscribeService]}
  screenService={[Function ScreenService]}
>
  <epgu-constructor-screen-container>
    <div
      class="screen-body"
    >
      <epgu-constructor-page-name>
        
      </epgu-constructor-page-name>
      <epgu-constructor-screen-pad
        class="body"
      >
        <epgu-constructor-components-list />
      </epgu-constructor-screen-pad>
    </div>
    <div
      class="screen-footer"
    >
      <lib-button
        class="btn__submit"
      >
         
      </lib-button>
    </div>
  </epgu-constructor-screen-container>
</epgu-constructor-custom-screen>
`;
